### PR TITLE
Add map with player origin

### LINKED ISSUE #221: Map with player origin

### PR DESCRIPTION
**Pull request by 1-alex98**
Fixes #221

### CODE REVIEW

**Comment by 1-alex98** [comment]
Somewhat like that 
grafik

File: public/js/app/world-map.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
+/**
+ * The following code inspired by https://www.amcharts.com/demos/rotating-globe-with-circles/
```
Comment (Geosearchef) [inline public/js/app/world-map.js]: is

File: public/js/app/world-map.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+ * @param data the list of players by country
+ */
+function createMap(data) {
+  am4core.ready(function () {
+
+    // Themes begin
+    am4core.useTheme(am4themes_animated);
+    // Themes end
```
Comment (Geosearchef) [inline public/js/app/world-map.js:10]: Remove both comments

File: public/js/app/world-map.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+    try {
+      chart.geodata = am4geodata_worldLow;
+    } catch (e) {
+      chart.raiseCriticalError(new Error("Map geodata could not be loaded. Please download the latest <a href=\"https://www.amcharts.com/download/download-v4/\">amcharts geodata</a> and extract its contents into the same directory as your amCharts files."));
+    }
+
+
+    var label = chart.createChild(am4core.Label);
+    label.text = "Players by country currently online";
```
Comment (Geosearchef) [inline public/js/app/world-map.js]: Players currently online by country

File: public/js/app/world-map.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    label.align = "left";
+    label.valign = "bottom";
+    label.fill = am4core.color("#927459");
+    label.background = new am4core.RoundedRectangle();
+    label.background.cornerRadius(10, 10, 10, 10);
+    label.padding(10, 10, 10, 10);
+    label.marginLeft = 30;
+    label.marginBottom = 30;
+    label.background.strokeOpacity = 0.3;
+    label.background.stroke = am4core.color("#927459");
```
Comment (Geosearchef) [inline public/js/app/world-map.js:34]: Same color as above, extract constant.

File: public/js/app/world-map.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    dataSource.text = "Data source: FAF lobby server";
+    dataSource.fontSize = 12;
+    dataSource.align = "left";
+    dataSource.valign = "top";
+    dataSource.fill = am4core.color("#927459");
+    dataSource.padding(10, 10, 10, 10);
+    dataSource.marginLeft = 30;
+    dataSource.marginTop = 30;
+
+// Set projection
```
Comment (Geosearchef) [inline public/js/app/world-map.js]: Indent

File: public/js/app/world-map.js [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    chart.projection = new am4maps.projections.Orthographic();
+    chart.panBehavior = "rotateLongLat";
+    chart.padding(20, 20, 20, 20);
+
+// Add zoom control
```
Comment (Geosearchef) [inline public/js/app/world-map.js]: Indent

File: public/js/app/world-map.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    chart.zoomControl = new am4maps.ZoomControl();
+
+    var homeButton = new am4core.Button();
+    homeButton.events.on("hit", function () {
```
Comment (Geosearchef) [inline public/js/app/world-map.js:57]: Replace with arrow function? Do we already support ECMAScript 6?

File: public/js/app/world-map.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    homeButton.parent = chart.zoomControl;
+    homeButton.insertBefore(chart.zoomControl.plusButton);
+
+    chart.backgroundSeries.mapPolygons.template.polygon.fill = am4core.color("#bfa58d");
+    chart.backgroundSeries.mapPolygons.template.polygon.fillOpacity = 1;
+    chart.deltaLongitude = 20;
+    chart.deltaLatitude = -20;
+
+// limits vertical rotation
+    chart.adapter.add("deltaLatitude", function (delatLatitude) {
```
Comment (Geosearchef) [inline public/js/app/world-map.js:75]: Arrow fn?

File: public/js/app/world-map.js [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+
+    var graticuleSeries = chart.series.push(new am4maps.GraticuleSeries());
+    graticuleSeries.mapLines.template.stroke = am4core.color("#fff");
+    graticuleSeries.fitExtent = false;
+    graticuleSeries.mapLines.template.strokeOpacity = 0.2;
+    graticuleSeries.mapLines.template.stroke = am4core.color("#fff");
+
+
+    var measelsSeries = chart.series.push(new am4maps.MapPolygonSeries());
```
Comment (Geosearchef) [inline public/js/app/world-map.js:141]: Why that variable naming?

File: public/js/app/world-map.js [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    measelsSeries.tooltip.dy = -5;
+
+    var measelTemplate = measelsSeries.mapPolygons.template;
+    measelTemplate.fill = am4core.color("#bf7569");
+    measelTemplate.strokeOpacity = 0;
+    measelTemplate.fillOpacity = 0.75;
+    measelTemplate.tooltipPosition = "fixed";
+
+
+    var hs2 = measelsSeries.mapPolygons.template.states.create("hover");
```
Comment (Geosearchef) [inline public/js/app/world-map.js:155]: hs2?

File: public/js/app/world-map.js [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    hs2.properties.fill = am4core.color("#86240c");
+
+    polygonSeries.events.on("inited", function () {
+      polygonSeries.mapPolygons.each(function (mapPolygon) {
+        var count = data[mapPolygon.id];
+
+        if (count > 0) {
+          var polygon = measelsSeries.mapPolygons.create();
+          polygon.multiPolygon = am4maps.getCircle(mapPolygon.visualLongitude, mapPolygon.visualLatitude, Math.max(0.2, Math.log(count) * Math.LN10 / 10));
+          polygon.tooltipText = mapPolygon.dataItem.dataContext.name + ": " + count;
```
Comment (Geosearchef) [inline public/js/app/world-map.js:166]: String format? I remember sth in my head about js not being capable of thattthough. Doesn't warrent a helper funct.

**Review by Geosearchef — COMMENTED** [review]
Nice work!

**Comment by 1-alex98** [comment]
The code was copied from the link in the comment. I do not plan to change it in anyway. I fixed the indent, the typos and Strings. I want to to stay as close to the original as possible to should it chnage it stays mantainable. 

Btw support for arrow functions depends on the browser.

File: routes/lobby_api.js [hunk 12]
```diff
@@ -4,34 +4,53 @@ let isFetching = false;
...
+exports = module.exports = function (req, res) {
+  let resource = req.query.resource;
+  if (cache[resource]) {
+    if (isFetching || Date.now() - cache[resource].pollTime < PLAYER_COUNT_UPDATE_INTERVAL) {
+      return res.send(cache[resource].data);
+    }
+  }
+  isFetching = true;
+  let queryResource = resource;
+  if (resource === "countries") {
```
Comment (1-alex98) [inline routes/lobby_api.js:16]: if it is countries ask for players and do a mapping (done later in the code)

File: routes/lobby_api.js [hunk 13]
```diff
@@ -4,34 +4,53 @@ let isFetching = false;
...
-    request(process.env.LOBBY_API_URL + "/" + resource, function (error, response, body) {
-        let data = [];
 
-        if (body) {
-            data = JSON.parse(body);
+    cache[queryResource] = {
+      pollTime: Date.now(),
+      data: data.length.toString()
+    };
+    if (queryResource === "players") {
```
Comment (axel1200) [inline routes/lobby_api.js]: if we asked for players on the lobby server also calculate/map the countries resource.

**Comment by 1-alex98** [comment]
@Geosearchef dont wanna approve it?

**Comment by 1-alex98** [comment]
so sad nobody looks at it

**Review by Geosearchef — APPROVED** [review]

**Comment by 1-alex98** [comment]
to sad you dont have write access hence your review does not count Xd

**Comment by 1-alex98** [comment]
@Brutus5000 needs to approve

**Review by Brutus5000 — APPROVED** [review]

**Pull request merged**